apps/emqx_utils/src/emqx_utils_uri.erl

```diff
@@ -0,0 +1,392 @@
+port(#{authority := undefined}) -> undefined;
+port(#{authority := #{port := Port}}) -> Port.
+
+-spec path(uri()) -> path().
+path(#{path := Path}) -> Path.
+
+-spec query(uri()) -> emqx_maybe:t(query()).
+query(#{query := Query}) -> Query.
+
+-spec fragment(uri()) -> emqx_maybe:t(fragment()).
+fragment(#{fragment := Fragment}) -> Fragment.
+
+-spec request_base(uri()) -> {ok, request_base()} | {error, term()}.
+request_base(URI) when is_map(URI) ->
+    case emqx_http_lib:uri_parse(iolist_to_binary(base_url(URI))) of
+        {error, Reason} -> {error, Reason};
+        {ok, URIMap} -> {ok, maps:with([scheme, host, port], URIMap)}
+    end;
+request_base(URIString) when is_list(URIString) orelse is_binary(URIString) ->
+    request_base(parse(URIString)).
+
+%%--------------------------------------------------------------------
+%% Helper functions
+%%--------------------------------------------------------------------
+
+parse_scheme(<<>>) -> undefined;
+parse_scheme(Scheme) -> Scheme.
+
+parse_query(<<>>) -> undefined;
+parse_query(<<$?, Query/binary>>) -> Query.
+
+parse_fragment(<<>>) -> undefined;
+parse_fragment(<<$#, Fragment/binary>>) -> Fragment.
+
+-define(AUTHORITY_REGEX,
+    ("^(?<userinfo>.*@)?"
+    "(?:(?:\\[(?<host_ipv6>[a-z\\d\\.:]*)\\])|(?<host_regular>[^:]*?)|(?<host_loose>.*?))"
+    "(?<port>:\\d+)?$")
+).
+
+authority_regexp() ->
+    {ok, RE} = re:compile(?AUTHORITY_REGEX, [caseless]),
+    RE.
+
+parse_authority(<<>>) ->
+    undefined;
+parse_authority(<<$/, $/, Authority/binary>>) ->
+    %% Authority regexp always matches
+    {match, [UserInfoMatch, HostIPv6, HostRegular, HostLoose, PortMatch]} = re:run(
+        Authority, authority_regexp(), [
+            {capture, [userinfo, host_ipv6, host_regular, host_loose, port], binary}
+        ]
+    ),
+    UserInfo = parse_userinfo(UserInfoMatch),
+    {HostType, Host} = parse_host(HostIPv6, HostRegular, HostLoose),
+    Port = parse_port(PortMatch),
+    #{
+        userinfo => UserInfo,
+        host => Host,
+        host_type => HostType,
+        port => Port
+    }.
+
+parse_userinfo(<<>>) -> undefined;
+parse_userinfo(UserInfoMatch) -> binary:part(UserInfoMatch, 0, byte_size(UserInfoMatch) - 1).
+
+parse_host(<<>>, <<>>, Host) -> {loose, Host};
+parse_host(<<>>, Host, <<>>) -> {regular, Host};
+parse_host(Host, <<>>, <<>>) -> {ipv6, Host}.
+
+parse_port(<<>>) -> undefined;
+parse_port(<<$:, Port/binary>>) -> binary_to_integer(Port).
+
+%% https://datatracker.ietf.org/doc/html/rfc3986#appendix-B
+%%
+%% > The following line is the regular expression for breaking-down a
+%% > well-formed URI reference into its components.
+%%
+%% > ^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?
+%%
+%% We skip capturing some unused parts of the regex.
+
+-define(URI_REGEX,
+    ("^(?:(?<scheme>[^:/?#]+):)?(?<authority>//[^/?#]*)?"
+    "(?<path>[^?#]*)(?<query>\\?[^#]*)?(?<fragment>#.*)?")
+).
+
+uri_regexp() ->
+    {ok, RE} = re:compile(?URI_REGEX, [caseless]),
```

Comment:
maybe worth caching it in persistent term in a on_load callback.
simple profiling shows re:compile is 20x as persistent_term:get

Reply:
✅ 
Currently, we do not parse URIs anywhere in hot loops only on resource/auth creation.
But let the code be a bit more foolproof and efficient.